homeassistant/components/light/limitlessled.py

```diff
@@ -8,171 +8,273 @@
-    'yellow_orange': [0xDA, 0xA5, 0x20],
-    'orange': [0xFF, 0xA5, 0x00],
-    'red': [0xFF, 0x00, 0x00],
-    'pink': [0xFF, 0xC0, 0xCB],
-    'fusia': [0xFF, 0x00, 0xFF],
-    'lilac': [0xDA, 0x70, 0xD6],
-    'lavendar': [0xE6, 0xE6, 0xFA],
-}
-
-
-def _distance_squared(rgb1, rgb2):
-    """ Return sum of squared distances of each color part. """
-    return sum((val1-val2)**2 for val1, val2 in zip(rgb1, rgb2))
-
-
-def _rgb_to_led_color(rgb_color):
-    """ Convert an RGB color to the closest color string and color. """
-    return sorted((_distance_squared(rgb_color, color), name)
-                  for name, color in COLOR_TABLE.items())[0][1]
-
-
-def setup_platform(hass, config, add_devices_callback, discovery_info=None):
-    """ Gets the LimitlessLED lights. """
-    import ledcontroller
+from limitlessled import Color
+from limitlessled.bridge import Bridge
+from limitlessled.group.rgbw import RgbwGroup
+from limitlessled.group.white import WhiteGroup
+from limitlessled.pipeline import Pipeline
+from limitlessled.presets import COLORLOOP
 
-    # Handle old configuration format:
-    bridges = config.get('bridges', [config])
 
-    for bridge_id, bridge in enumerate(bridges):
-        bridge['id'] = bridge_id
+_LOGGER = logging.getLogger(__name__)
+REQUIREMENTS = ['limitlessled==1.0.0']
+RGB_BOUNDARY = 40
+DEFAULT_TRANSITION = 0
+DEFAULT_PORT = 8899
+DEFAULT_VERSION = 5
+DEFAULT_LED_TYPE = 'rgbw'
+WHITE = [255, 255, 255]
 
-    pool = ledcontroller.LedControllerPool([x['host'] for x in bridges])
 
+def legacy_setup(config, add_devices_callback):
+    """ Perform setup using legacy format. """
+    bridges = config.get('bridges', [config])
     lights = []
-    for bridge in bridges:
+    for bridge_conf in bridges:
+        bridge = Bridge(bridge_conf.get('host'))
         for i in range(1, 5):
             name_key = 'group_%d_name' % i
-            if name_key in bridge:
-                group_type = bridge.get('group_%d_type' % i, 'rgbw')
-                lights.append(LimitlessLED.factory(pool, bridge['id'], i,
-                                                   bridge[name_key],
-                                                   group_type))
-
+            if name_key in bridge_conf:
+                group_type = bridge_conf.get('group_%d_type' % i,
+                                             DEFAULT_LED_TYPE)
+                group = bridge.add_group(i,
+                                         bridge_conf.get(name_key),
+                                         group_type)
+                lights.append(LimitlessLEDGroup.factory(group))
     add_devices_callback(lights)
 
 
-class LimitlessLED(Light):
-    """ Represents a LimitlessLED light """
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """ Gets the LimitlessLED lights. """
+
+    # Two legacy configuration formats are supported to
+    # maintain backwards compatibility.
+    legacy_setup(config, add_devices_callback)
```

Comment:
The automation component does it on the fly, not rewrite
configuration.yaml. So it takes in a legacy dict, emits a warwning and
converts it to a new config format dict.

On Fri, Nov 27, 2015 at 2:04 PM, happyleavesaoc <notifications@github.com>
wrote:

> In homeassistant/components/light/limitlessled.py
> <https://github.com/balloob/home-assistant/pull/671#discussion_r46071492>:
>
> >      add_devices_callback(lights)
> >
> >
> > -class LimitlessLED(Light):
> > -    """ Represents a LimitlessLED light """
> > +def setup_platform(hass, config, add_devices_callback, discovery_info=None):
> > +    """ Gets the LimitlessLED lights. """
> > +
> > +    # Two legacy configuration formats are supported to
> > +    # maintain backwards compatibility.
> > +    legacy_setup(config, add_devices_callback)
>
> The legacy configuration uses a different set of keys. The component just
> tries all three formats. So technically you could mix-and-match within your
> configuration.
>
> It would be possible to re-write the configuration, but do we want that
> behavior? Is there precedent for that with other components?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/671/files#r46071492>.
>



-- 
PaulusSchoutsen.nl
It's nice to be important but it's more important to be nice.